barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp

```diff
@@ -14,114 +15,203 @@ template class DSLBigInts<MegaCircuitBuilder>;
 
 template <typename Builder>
 void build_constraints(Builder& builder,
-                       AcirFormat const& constraint_system,
+                       AcirFormat& constraint_system,
                        bool has_valid_witness_assignments,
                        bool honk_recursion)
 {
+    constraint_system.gates_per_opcode = std::vector<size_t>(constraint_system.num_acir_opcodes);
+    size_t prev_gate_count = 0;
+
+    auto compute_gate_diff = [&]() {
+        size_t new_gate_count = builder.get_num_gates();
+        size_t diff = new_gate_count - prev_gate_count;
+        prev_gate_count = new_gate_count;
+        return diff;
+    };
+
     // Add arithmetic gates
-    for (const auto& constraint : constraint_system.poly_triple_constraints) {
+    for (size_t i = 0; i < constraint_system.poly_triple_constraints.size(); ++i) {
+        const auto& constraint = constraint_system.poly_triple_constraints[i];
         builder.create_poly_gate(constraint);
```

Comment:
Do we know how much longer constraint building takes with this?

Reply:
Agreed, seems like we're going to do a lot of extra work on any interaction with the backend which is going to be unnecessary a lot of the time.

Reply:
I can probably switch it on with a boolean flag that is only active in the gate count command

Reply:
I don't know if this change makes sense, doing this small bit of math isn't likely to be slow. I wasn't sure it needed to be conditional, it would depend on knowing how much longer it takes

Reply:
Talked to Alvaro, this was heavier than I realized but we still did a pass here :)